include/sampleflow/producers/differential_evaluation_mh.h

```diff
@@ -0,0 +1,201 @@
+
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 by the SampleFlow authors.
+//
+// This file is part of the SampleFlow library.
+//
+// The SampleFlow library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of SampleFlow.
+//
+// ---------------------------------------------------------------------
+
+#ifndef SAMPLEFLOW_PRODUCERS_DIFFERENTIAL_EVALUATION_MH_H
+#define SAMPLEFLOW_PRODUCERS_DIFFERENTIAL_EVALUATION_MH_H
+
+#include <sampleflow/producer.h>
+#include <sampleflow/scope_exit.h>
+#include <sampleflow/types.h>
+
+#include <algorithm>
+#include <random>
+#include <cmath>
+#include <valarray>
+
+namespace SampleFlow
+{
+  namespace Producers
+  {
+    /**
+     * An implementation of Differential Evaluation Metropolis Hastings.
+     * This producer functions very similarly to the regular Metropolis Hastings
+     * producer - consult the documentation of that class for more details.
+     * This class provides the same functionality as Metropolis Hastings but does
+     * so over N chains simultaneously, which can improve the rate of convergence.
+     */
+    template <typename OutputType>
+    class DifferentialEvaluationMetropolisHastings : public Producer<OutputType>
+    {
+      public:
+        /**
+         * The principal function of this class. Starting from the given
+         * initial samples $x_{i, 0}$, it produces a multiple chains of
+         * samples $x_{i, k}$ that are passed through the signal of the
+         * base class to Consumer objects.
+         *
+         * @param[in] starting_points The initial samples of $x_{i, 0}$ for
+         *   each chain {i}. Must have at least 3 elements.
+         * @param[in] log_likelihood A function object that, when called
+         *   with a sample $x$, returns $\log(\pi(x))$, i.e., the natural
+         *   logarithm of the likelihood function evaluated at the sample.
+         * @param[in] perturb A function object that, when given a sample
+         *   $x$ as well as a vector of rejected samples $y$,
+         *   returns a pair of values containing the following:
+         *   <ol>
+         *   <li> A different sample $\tilde x$ that is perturbed
+         *     in some way from the given sample $x$.
+         *   <li> The relative probability of the transition $x\to\tilde x$
```

Comment:
This probably has to be updated based on what one needs in the algorithm. Look it up in the original paper.

Reply:
@bangerth Are you referring to the the second item in the pair? I think it is the same as in regular metropolis hastings - the acceptance ratio is calculated in exactly the same way.

Reply:
forget this -- I was commenting on the DR algorithm.